From 8b35519c8e910407bfc7c64cd8d321035f140e91 Mon Sep 17 00:00:00 2001
From: Josh Wilsdon <jwilsdon@joyent.com>
Date: Tue, 31 Jan 2017 22:52:55 -0800
Subject: [PATCH] AGENT-1057 log setup for agents should move to a -setup
 service instead of the agent's startup script Reviewed by: Orlando Vazquez
 <orlando@joyent.com> Reviewed by: Trent Mick <trentm@gmail.com> Approved by:
 Orlando Vazquez <orlando@joyent.com>

---
 npm/postinstall.sh                         | 23 ++++-----
 smf/manifests/cn-agent-setup.xml.in        | 55 ++++++++++++++++++++++
 smf/manifests/cn-agent.xml.in              |  6 +--
 smf/method/{cn-agent.in => cn-agent-setup} |  6 +--
 4 files changed, 73 insertions(+), 17 deletions(-)
 create mode 100644 smf/manifests/cn-agent-setup.xml.in
 rename smf/method/{cn-agent.in => cn-agent-setup} (91%)

diff --git a/npm/postinstall.sh b/npm/postinstall.sh
index 7bcf8e5..9ffb2e7 100755
--- a/npm/postinstall.sh
+++ b/npm/postinstall.sh
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright 2015 Joyent, Inc.
+# Copyright 2017 Joyent, Inc.
 #
 
 if [[ "${SDC_AGENT_SKIP_LIFECYCLE:-no}" = "yes" ]]; then
@@ -67,6 +67,10 @@ function subfile
         file_port='5310'
         file_enabled='false'
         ;;
+    setup)
+        file_enabled='true'
+        file_port='0'
+        ;;
     *)
         fatal 'Unknown agent type: "%s".' "${agent_type}"
         ;;
@@ -88,16 +92,13 @@ function subfile
 #
 function import_smf_manifest
 {
-    local method_in="$ROOT/smf/method/$AGENT.in"
-    local method_out="$ROOT/smf/method/$AGENT"
     local manifest0_in="$ROOT/smf/manifests/$AGENT.xml.in"
     local manifest0_out="$SMF_DIR/$AGENT.xml"
     local manifest1_in="$ROOT/smf/manifests/$AGENT-update.xml.in"
     local manifest1_out="$SMF_DIR/$AGENT-update.xml"
+    local manifest2_in="$ROOT/smf/manifests/${AGENT}-setup.xml.in"
+    local manifest2_out="$SMF_DIR/${AGENT}-setup.xml"
 
-    if [[ ! -f "${method_in}" ]]; then
-        fatal 'could not find smf method input file: %s' "${method_in}"
-    fi
     if [[ ! -f "${manifest0_in}" ]]; then
         fatal 'could not find smf manifest input file: %s' "${manifest0_in}"
     fi
@@ -105,11 +106,6 @@ function import_smf_manifest
         fatal 'could not find smf manifest input file: %s' "${manifest1_in}"
     fi
 
-    if ! subfile "${method_in}" "${method_out}" 'normal' ||
-      ! chmod +x "${method_out}"; then
-        fatal 'could not process smf method (%s)' "${method_in}"
-    fi
-
     if ! subfile "${manifest0_in}" "${manifest0_out}" 'normal' ||
       ! svccfg import "${manifest0_out}"; then
         fatal 'could not process smf manifest (%s)' "${manifest0_in}"
@@ -119,6 +115,11 @@ function import_smf_manifest
       ! svccfg import "${manifest1_out}"; then
         fatal 'could not process smf manifest (%s)' "${manifest1_in}"
     fi
+
+    if ! subfile "${manifest2_in}" "${manifest2_out}" 'setup' ||
+      ! svccfg import "${manifest2_out}"; then
+        fatal 'could not process smf manifest (%s)' "${manifest2_in}"
+    fi
 }
 
 #
diff --git a/smf/manifests/cn-agent-setup.xml.in b/smf/manifests/cn-agent-setup.xml.in
new file mode 100644
index 0000000..1943587
--- /dev/null
+++ b/smf/manifests/cn-agent-setup.xml.in
@@ -0,0 +1,55 @@
+<?xml version="1.0"?>
+<!DOCTYPE service_bundle SYSTEM "/usr/share/lib/xml/dtd/service_bundle.dtd.1">
+<!--
+    This Source Code Form is subject to the terms of the Mozilla Public
+    License, v. 2.0. If a copy of the MPL was not distributed with this
+    file, You can obtain one at http://mozilla.org/MPL/2.0/.
+-->
+
+<!--
+    Copyright (c) 2017, Joyent, Inc.
+-->
+
+<service_bundle type="manifest" name="cn-agent-setup">
+  <service name="smartdc/agent/cn-agent-setup" type="service" version="@@VERSION@@">
+
+    <create_default_instance enabled="@@ENABLED@@"/>
+    <single_instance/>
+
+    <dependency name="network" grouping="require_all" restart_on="error" type="service">
+      <service_fmri value="svc:/milestone/network:default"/>
+    </dependency>
+
+    <exec_method
+      type="method"
+      name="start"
+      exec="@@ROOT@@/smf/method/cn-agent-setup %m"
+      timeout_seconds="600">
+      <method_context>
+        <method_credential user="root" group="staff"/>
+        <method_environment>
+          <envvar name="PATH" value="/usr/bin:/usr/sbin"/>
+        </method_environment>
+      </method_context>
+    </exec_method>
+
+    <exec_method type="method" name="stop" exec=":true" timeout_seconds="60" />
+
+    <property_group name="startd" type="framework">
+      <propval name="duration" type="astring" value="transient" />
+      <propval name="ignore_error" type="astring" value="core,signal" />
+    </property_group>
+
+    <property_group name="application" type="application">
+    </property_group>
+
+    <stability value="Evolving"/>
+
+    <template>
+      <common_name>
+        <loctext xml:lang="C">SmartDataCentre Compute Node Agent Setup</loctext>
+      </common_name>
+    </template>
+
+  </service>
+</service_bundle>
diff --git a/smf/manifests/cn-agent.xml.in b/smf/manifests/cn-agent.xml.in
index d0624ef..ab405f8 100644
--- a/smf/manifests/cn-agent.xml.in
+++ b/smf/manifests/cn-agent.xml.in
@@ -7,7 +7,7 @@
 -->
 
 <!--
-    Copyright (c) 2015, Joyent, Inc.
+    Copyright (c) 2017, Joyent, Inc.
 -->
 
 <service_bundle type="manifest" name="cn-agent">
@@ -27,12 +27,12 @@
     <exec_method
       type="method"
       name="start"
-      exec="@@ROOT@@/smf/method/cn-agent %m"
+      exec="/usr/bin/ctrun -l child -o noorphan @@ROOT@@/node/bin/node --abort_on_uncaught_exception @@ROOT@@/bin/cn-agent &amp;"
       timeout_seconds="60">
       <method_context>
         <method_credential user="root" group="staff"/>
         <method_environment>
-          <envvar name="PATH" value="@@PREFIX@@/bin:/usr/local/bin:/opt/local/bin:/usr/bin:/usr/sbin:/bin"/>
+          <envvar name="PATH" value="@@PREFIX@@/bin:/usr/bin:/usr/sbin"/>
           <envvar name="PORT" value="@@PORT@@"/>
         </method_environment>
       </method_context>
diff --git a/smf/method/cn-agent.in b/smf/method/cn-agent-setup
similarity index 91%
rename from smf/method/cn-agent.in
rename to smf/method/cn-agent-setup
index 72cfac9..bc87fd2 100755
--- a/smf/method/cn-agent.in
+++ b/smf/method/cn-agent-setup
@@ -6,11 +6,12 @@
 #
 
 #
-# Copyright (c) 2015, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 #
-# SMF Startup script for cn-agent
+# Runs on node (CN + HN) boot to setup log rotation for cn-agent logs.
+#
 
 set -o xtrace
 
@@ -54,7 +55,6 @@ function setup_logadm {
 case "$1" in
 'start')
     setup_logadm
-    /usr/bin/ctrun -l child -o noorphan @@ROOT@@/node/bin/node --abort_on_uncaught_exception @@ROOT@@/bin/cn-agent &
     ;;
 
 'stop')
-- 
2.21.0

